File: problems/2303-calculate-amount-paid-in-taxes/solution.cpp
class Solution {
public:
    double calculateTax(vector<vector<int>>& brackets, int income) {
        double tax = 0.0;
        int prev_upper = 0;
        for (int i = 0; i < brackets.size(); ++i) {
            int upper = brackets[i][0];
            int percent = brackets[i][1];
            int taxable = min(income, upper) - prev_upper;
            if (taxable > 0) {
                tax += (double)taxable * (double)percent / 100.0;
            }
            prev_upper = upper;
            if (income <= upper) {
                break;
            }
        }
        return tax;
    }
};

File: problems/2303-calculate-amount-paid-in-taxes/solution.py
class Solution:
    def calculateTax(self, brackets: List[List[int]], income: int) -> float:
        tax = 0.0
        prev_upper = 0
        for upper, percent in brackets:
            if income <= 0:
                break
            taxable = min(income, upper - prev_upper)
            tax += taxable * percent / 100.0
            income -= taxable
            prev_upper = upper
        return tax